pkg/edvprovider/couchdbedvprovider/couchdbedvprovider.go

```diff
@@ -0,0 +1,176 @@
+	encryptedDocID string) error {
+	mapDocument := couchDBIndexMappingDocument{
+		IndexName:              indexedAttributeName,
+		MatchingEncryptedDocID: encryptedDocID,
+	}
+
+	documentBytes, err := json.Marshal(mapDocument)
+	if err != nil {
+		return err
+	}
+
+	err = store.Put(encryptedDocID+"_mapping_"+uuid.New().String(), documentBytes)
+	if err != nil {
+		return err
+	}
+
+	createIndexRequest := storage.CreateIndexRequest{
+		IndexStorageLocation: "EDV_EncryptedIndexesDesignDoc",
+		IndexName:            "EDV_IndexName",
+		WhatToIndex:          `{"fields": ["` + indexedField + `"]}`,
+	}
+
+	return store.CreateIndex(createIndexRequest)
+}
+
+// EncryptedIndexQuery does an EDV encrypted index query.
+// Due to the document mapping workaround described in CreateIndexOnSingleField, we first need to get
+// the "mapping document" and then use the ID we get from that to lookup the associated encrypted document.
+// Then we check that encrypted document to see if the value matches what was specified.
+func (c CouchDBEDVProvider) EncryptedIndexQuery(store storage.Store, query *models.Query) ([]string, error) {
+	idsOfDocsWithMatchingQueryIndexName, err := findDocsMatchingQueryIndexName(store, query.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	return findDocsMatchingQueryValue(idsOfDocsWithMatchingQueryIndexName, store, query)
+}
+
+func findDocsMatchingQueryIndexName(store storage.Store, queryIndexName string) (map[string]struct{}, error) {
+	itr, err := store.Query(`{
+		   "selector": {
+		       "` + indexedField + `": "` + queryIndexName + `"
+		   },
+			"use_index": ["EDV_EncryptedIndexesDesignDoc", "EDV_IndexName"]
+		}`)
+	if err != nil {
+		return nil, err
+	}
+
+	ok, err := itr.Next()
+	if err != nil {
+		return nil, err
+	}
+
+	// I really just want an array that doesn't allow for duplicates, but in golang a map is the closest I can get.
+	// The "struct{}" value is not used and should not occupy any additional space.
+	idsOfDocsWithAMatchingIndex := make(map[string]struct{})
+
+	for ok {
+		value, err := itr.Value()
+		if err != nil {
+			return nil, err
+		}
+
+		receivedCouchDBIndexMappingDocument := couchDBIndexMappingDocument{}
+
+		err = json.Unmarshal(value, &receivedCouchDBIndexMappingDocument)
+		if err != nil {
+			return nil, err
+		}
+
+		idsOfDocsWithAMatchingIndex[receivedCouchDBIndexMappingDocument.MatchingEncryptedDocID] = struct{}{}
+
+		ok, err = itr.Next()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return idsOfDocsWithAMatchingIndex, nil
+}
+
+func findDocsMatchingQueryValue(idsOfDocsWithAMatchingIndex map[string]struct{}, store storage.Store,
+	query *models.Query) ([]string, error) {
+	idsOfDocsWithAMatchingIndexAndValue := make([]string, 0)
+
+	for docID := range idsOfDocsWithAMatchingIndex {
+		documentBytes, err := store.Get(docID)
+		if err == storage.ErrValueNotFound {
+			return nil, edverrors.ErrDocumentNotFound
+		} else if err != nil {
```

Comment:
Don't need an 'else' since there's already a return.

Reply:
I don't see an 'else' here, so I don't understand what you mean. Or perhaps I misunderstood your comment - can you elaborate?

Reply:
if err == storage.ErrValueNotFound {
	return nil, edverrors.ErrDocumentNotFound
} else if err != nil {
. . .
}

Should be:

if err == storage.ErrValueNotFound {
	return nil, edverrors.ErrDocumentNotFound
}

if err != nil {
. . .
}

Reply:
They look equivalent for me. Is there a reason you prefer the second form? Note that I have an identical small chunk of code in operations.go, around line 288 that uses the same style.

(side-note: I considered moving these to a common place, but I had trouble figuring out the best place to put it and an appropriate method name, so I left it for now)

Reply:
I'm surprised that the linter allowed this code to pass. If you return from a function then there's no reason to perform an 'else'.

Reply:
This is an 'else if' not an 'else'.

Reply:
> This is an 'else if' not an 'else'.

So I would do:

if err != nil {
  if err == storage.ErrValueNotFound {
    return nil, edverrors.ErrDocumentNotFound
  }
  return nil, err
}

Reply:
Done! I also updated the other place where I have that same code snippet.